app/src/main/java/ch/epfl/sweng/rps/services/MatchmakingService.kt

```diff
@@ -0,0 +1,81 @@
+package ch.epfl.sweng.rps.services
+
+import android.util.Log
+import ch.epfl.sweng.rps.models.Game
+import com.google.firebase.functions.ktx.functions
+import com.google.firebase.ktx.Firebase
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.tasks.await
+
+class MatchmakingService {
+    private val cloudFunctions = CloudFunctions()
+
+    fun queue(gameMode: Game.GameMode): Flow<QueueStatus> = flow {
+        Log.i("MatchmakingService", "Queueing for game mode ${gameMode.toGameModeString()}")
+        emit(QueueStatus.Queued)
+        Log.i("MatchmakingService", "Sending request to cloud function")
+        val gameId = cloudFunctions.queue(gameMode)
+        Log.i("MatchmakingService", "Received game id $gameId")
+
+        val service = ServiceLocator.getInstance().getGameServiceForGame(gameId)
+        Log.i("MatchmakingService", "Got service $service")
+        emit(QueueStatus.Accepted(service))
+    }
+
+    suspend fun acceptInvitation(invitationId: String): FirebaseGameService {
+        val gameId = cloudFunctions.acceptInvitation(invitationId)
+        return ServiceLocator.getInstance().getGameServiceForGame(gameId)
+    }
+
+    suspend fun currentGame(): FirebaseGameService? {
+        val repo = ServiceLocator.getInstance().repository
+        val games = repo.gamesOfUser(repo.getCurrentUid())
+        if (games.isEmpty()) {
+            return null
+        }
+        if (games.size > 1) {
+            Log.e(
+                "MatchmakingService",
+                "More than one game for user ${repo.getCurrentUid()}: ${games.map { it.id }}"
+            )
+        }
+        return ServiceLocator.getInstance().getGameServiceForGame(games[0].id)
+    }
+
+    sealed class QueueStatus {
+        object Queued : QueueStatus()
+        class Accepted(val gameService: GameService) : QueueStatus()
+    }
```

Comment:
Could you elaborate on when this class would be used? What is the reason to make it sealed?

Reply:
[Sealed classes﻿](https://kotlinlang.org/docs/sealed-classes.html) allow to use these as [Union](https://en.wikipedia.org/wiki/Union_type) classes.

Reply:
It's used to have a status on the current state of the queue. Read more about [flows](https://kotlinlang.org/docs/flow.html) if you need.